asciidoc/courses/llm-fundamentals/modules/1-introduction/lessons/2-hallucination/lesson.adoc

```diff
@@ -0,0 +1,129 @@
+
+For instance, if an LLM's last update was in September 2022, it would be unaware of world events or advancements in various fields that occurred post that date, leading to potential gaps in its knowledge or responses that seem out of touch with current realities.
+
+If the user is asking a question on information that is hard to find, or outside of the public domain, it will be next to impossible for an LLM to provide an accurate response.
+
+Luckily, this is where a knowledge graph comes in.
+
+
+// How?
+== Improving LLM Accuracy
+
+The following methods can be employed to help guide LLMs to produce more consistent and accurate results.
+
+
+=== Prompt Engineering
+
+Prompt engineering is an iterative process of crafting specific and deliberate questions that guide the LLM toward the desired response.
+By refining the way questions are posed, developers can achieve better results from existing models without the need for retraining.
+
+For example, if you require a summary of a blog post, rather than asking _"What is this book about?"_, a more appropriate response would be _"Provide a concise, three-sentence summary and three tags for this blog post."_
+
+You could even throw in _"Return the response as JSON"_ and provide an example output to make it easier to parse in the programming language of your choice.
+
+[TIP]
+.Be Positive
+====
+When writing a prompt, I aim to provide positive instructions.
+For the same reason a toddler will ignore you telling them not to draw on the walls, the model will act based on what you don't want in mind, rather than what you need.
+====
+// TODO: better example
+
+
+=== In-Context Learning
+
+In-context learning provides the model with a set of examples to inform its responses, helping it comprehend the task better.
+By presenting a series of relevant examples, the model can deliver more accurate answers, especially for niche or specialized tasks.
+
+For example, if the LLM starts talking about Tim Cook and iPhones every time you mention apples, you may want to add to the prompt that _apple_ refers to the fruit.
+
+
+=== Fine-Tuning
+
+Fine-tuning involves additional training of a language model on a smaller, task-specific dataset after its primary training phase. This approach allows developers to specialize the model for specific domains or tasks, enhancing its accuracy and relevance.
+
+For instance, fine-tuning an existing model on the fine nuances of your programming language would enhance its capability to write performant code.
+This method is by far the most complicated method, which involves both technical knowledge and domain expertise.
+
+A simpler approach would be to _ground_ the model by providing information with the prompt.
+
+
+=== Grounding
+
+Grounding allows a language model to reference external, up-to-date sources or databases to enrich its responses.
```

Comment:
this is too vague, how do you to this?
Need to say that those facts are retrieved based on the user question from the data-sources and added to the prompt as high priority, factual context for the LLM to use to answer the question. and that it's instructed to not use its training data.

Reply:
Grounding is in fact a variant of in-context / few-shot learning, only that the examples are not hard coded but are relevant information coming from the data-sources